src/machine.rs

```diff
@@ -789,3 +791,33 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'mir, 'tcx> {
         res
     }
 }
+
+#[derive(Clone)]
+pub struct CurrentSpan<'a, 'tcx, 'mir> {
+    span: Option<Span>,
+    machine: &'a Evaluator<'tcx, 'mir>,
+}
+
+impl<'a, 'tcx, 'mir> CurrentSpan<'a, 'tcx, 'mir> {
+    pub fn get(&mut self) -> rustc_span::Span {
+        if self.span.is_none() {
+            self.span = Some(self.current_span());
+        }
+        self.span.unwrap()
```

Comment:
I think `Option` has a nice combinator method that does this... ah, `get_or_insert`. Please use that, which will avoid the `unwrap`.